metronome/checkpointing/service/src/io/iohk/metronome/checkpointing/service/CheckpointingService.scala

```diff
@@ -60,14 +68,12 @@ class CheckpointingService[F[_]: Sync: Applicative, N](
     updateLedger(block).flatMap {
       case Some(ledger) =>
         if (block.hash != commitQC.blockHash)
-          saveState(block.header, ledger).as(true)
+          false.pure[F]
         else
           for {
-            //TODO: PM-3110:
-            // chkpOpt <- constructCheckpoint(ledger, commitQC)
-            _ <- saveState(block.header, ledger)
-            //TODO: PM-3110:
-            // _ <- chkpOpt.map(pushCheckpoint).getOrElse(().pure[F])
+            chkpOpt <- constructCheckpointCert(commitQC)
+            _       <- saveState(block.header, ledger)
+            -       <- chkpOpt.map(pushCheckpoint).getOrElse(().pure[F])
```

Comment:
If the app crashes at this point we will push the checkpoint on restart. Maybe we should have another marker for whether the push was completed? 🤔 

To be solved in PM-3137 anyway

Reply:
Right, because if we crash before returning from `executeBlock` then the `BlockExecutor` will not remember that this block was indeed executed. However it will wait for the next commit certificate to arrive, at which point there may be newer candidates to push. But if there aren't, you get a longer proof for the same candidate. 

This longer proof thing can happen to other nodes as well that somehow miss a Q.C. and construct a proof based on the next block. It's the malleability issue in a different disguise, and the interpreter has to be able to handle it.

So I wouldn't bother with the completion, especially since pushing the message doesn't mean it was received, it's very best effort.